docs/proposals/16-adminkubeconfig-subresource.md

```diff
@@ -0,0 +1,126 @@
+
+This `GEP` introduces new  `Shoot` subresource called `AdminKubeConfigRequest` allowing for users to dynamically generate a short-lived `kubeconfig` that can be used to access the `Shoot` cluster as `cluster-admin`.
+
+## Motivation
+
+Today, when access to the created `Shoot` clusters is needed, a `kubeconfig` with static token credentials is used. This static token is in the `system:masters` group, granting it `cluster-admin` privileges. The `kubeconfig` is generated when the cluster is reconciled, stored in `ShootState` and replicated in the `Project`'s namespace in a `Secret`. End-users can fetch the secret and use the `kubeconfig` inside it.
+
+There are several problems with this approach:
+
+- The token in the `kubeconfig` does not have any expiration, so end-users have to request a `kubeconfig` credential rotation if they want revoke the token.
+- There is no user identity in the token. e.g. if user `Joe` gets the `kubeconfig` from the `Secret`, user in that token would be `system:cluster-admin` and not `Joe` when accessing the `Shoot` cluster with it. This makes auditing events in the cluster almost impossible.
+
+### Goals
+
+- Add a `Shoot` subresource called `adminkubeconfig` that would produce a `kubeconfig` used to access that `Shoot` cluster.
+- The `kubeconfig` is not stored in the API Server, but generated for each request.
+- In the `AdminKubeConfigRequest` send to that subresource, end-users can specify the expiration time of the credential.
+- The identity (user) in the Gardener cluster would be part of the identity (x509 client certificate). E.g if `Joe` authenticates against the Gardener API server, the generated certificate for `Shoot` authentication would have the following subject:
+
+  - Common Name: `Joe`
+  - Organisation: `system:masters`
+
+- The maximum validity of the certificate can be enforced by setting a flag on the `gardener-apiserver`.
+- Deprecate and remove the old `{shoot-name}.kubeconfig` secrets in each `Project` namespace.
+
+### Non-Goals
+
+- Generate `OpenID Connect` kubeconfigs
+
+## Proposal
+
+The `gardener-apiserver` would serve a new `shoots/adminkubeconfig` resource. It can only accept `CREATE` calls and accept `AdminKubeConfigRequest`. A `AdminKubeConfigRequest` would have the following structure:
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: AdminKubeConfigRequest
+spec:
+  expirationSeconds: 3600
+```
+
+Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeConfigRequest` is configured by `--shoot-admin-kubeconfig-max-expiration` flag in the `gardener-apiserver`.
+
+When such request is received, the API server would find the `ShootState` associated with that cluster and generate a `kubeconfig`. The x509 client certificate would be signed by the `Shoot` cluster's CA and the user used in the subject's common name would be from the `User.Info` used to make the request.
```

Comment:
Do you have any performance concerns for generating x509 client certs in the API server? I can imagine, that this will become quite the heavy load on the API server once it's exposed to endusers and might become a scalability limitation.
I assume, you still chose the synchronous design here in order to not persist the client creds anywhere, right?
Still, I'm asking myself if we could make this work asynchronously to not run into any performance issues in the future.

Reply:
> Do you have any performance concerns for generating x509 client certs in the API server? 

I'm more concerned with the size of the `ShootState` that is all stored in the `kube-apiserver` in-memory. However, this is a separate concern.

> Still, I'm asking myself if we could make this work asynchronously to not run into any performance issues in the future.

The idea is to not store it anywhere. The API is very similar to the `TokenRequest` API in Kubernetes where also `JWT` tokens are generated on the flight.

We should probably introcude proper [flow controll and API fairness](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/), but this should be done in a separate PR.

Reply:
Makes sense, thanks for clarifying!